lantz_core/has_features.py

```diff
@@ -0,0 +1,789 @@
+                      }
+        feat_paras = {}                  # Sentinels changing feats behavior.
+
+        # List of the entries to remove from the class because they are
+        # destinated to a subpart.
+        to_remove = []
+
+        docs = dct.pop('_docs_') if '_docs_' in dct else None
+        print(docs)
+
+        # Set of seen subparts to avoid counting multiple times the same one
+        # which ahappens due to the context manager use.
+        seen_subparts = set()
+
+        # First we identify all elements in the passed dict to clean it up
+        # before creating the class.
+        for key, value in dct.iteritems():
+            if value is SUBPART_FUNC:
+                to_remove.append(key)
+
+            elif isinstance(value, _subpart):
+                if key in subparts:
+                    msg = 'Attempt to redeclare subpart {}'
+                    raise KeyError(msg.format(key))
+                if value not in seen_subparts:
+                    value._name_ = key
+                    subparts[key] = value
+                    seen_subparts.add(value)
+                elif len(key) > len(value._name_):
+                    subparts[key] = value
+                    del subparts[value._name_]
+                    to_remove.append(value._name_)
+                    value._name_ = key
+                else:
+                    to_remove.append(key)
+                value._aliases_.append(key)
+
+            elif isinstance(value, Feature):
+                feats[key] = value
+                value.name = key
+
+            elif isinstance(value, set_feat):
+                feat_paras[key] = value
+
+            elif isinstance(value, FunctionType):
+                if key.startswith(POST_GET_PREFIX):
+                    cust_feats['post_get'].append(key)
+                elif key.startswith(PRE_SET_PREFIX):
+                    cust_feats['pre_set'].append(key)
+                elif key.startswith(POST_SET_PREFIX):
+                    cust_feats['post_set'].append(key)
+                elif key.startswith(PRE_GET_PREFIX):
+                    cust_feats['pre_get'].append(key)
+                elif key.startswith(GET_PREFIX):
+                    cust_feats['get'].append(key)
+                elif key.startswith(SET_PREFIX):
+                    cust_feats['set'].append(key)
+
+        # Clean up class dictionary.
+        for k in chain(feat_paras, to_remove, subparts):
+            del dct[k]
+
+        # Create the class object.
+        cls = super(HasFeaturesMeta, meta).__new__(meta, name, bases, dct)
+
+        # Purge the list of base classes (for the most base class object is
+        # present)
+        bases = [b for b in bases if issubclass(b, AbstractHasFeatures)]
+
+        # Analyse the source code to find the doc for the defined Features.
+        # This will work as long as two subpart are not aliased in the same
+        # way which is probabbly good enough.
```

Comment:
I have never timed it. It happens only once when the class is created so it should not be too bad as long as people don't load the whole library of drivers at once. I like it as it is the natural way to document things in Sphinx so there is no surprise.